Add heartbeat and cool down period to instant-seal 💗

jimmychu0807 opened the pull request:
This PR is to address #6849, and has received immense support from @JoshOrndorff and @seunlanlege 

We add a new heartbeat stream that merge the transaction pool notification stream in together. It has three characteristics:

- you can use instant-seal as before (meaning no heartbeat at all) if you pass a `None` in the heartbeat option in `run_instant_seal` function.

- There is a `max_blocktime` param in the heartbeat option. If no transactions come within a `max_blocktime` period, a new block will be generated. Default set to 30 sec.

- There is a `min_blocktime` param in the heartbeat option. If multiple transactions come within a `min_blocktime` period since the last block is generated, it will wait for the `min_blocktime` to be passed since the last block import and then seal a new block. This is to prevent block being sealed too frequently.

Before merge:
- [x] Add documentation
- [x] Remove code from node template

JoshOrndorff commented on client/consensus/manual-seal/src/heartbeat_stream.rs:
```diff
@@ -0,0 +1,100 @@
+	}
+}
+
+pub struct HeartbeatStream<Hash> {
```
Needs Docs

JoshOrndorff commented on client/consensus/manual-seal/src/heartbeat_stream.rs:
```diff
@@ -0,0 +1,100 @@
+const DEFAULT_FINALIZE: bool = false;
+
+// ---
+pub struct HeartbeatOptions {
```
Needs docs

JoshOrndorff commented on client/consensus/manual-seal/src/heartbeat_stream.rs:
```diff
@@ -0,0 +1,100 @@
+		match hbs.pool_stream.poll_next_unpin(cx) {
+			Poll::Ready(Some(ec)) => {
+				// Ensure we are not producing block too frequently, based on `min_blocktime` setting
+				if Instant::now() - hbs.instant < Duration::from_secs(hbs.opts.min_blocktime) {
+					return Poll::Pending;
+				}
```
You've checked to make sure we don't author two blocks in short succession, but you haven't ensured that a second block will get authored in a reasonable amount of time. Imagine a node using instant seal. A user submits two transactions very close together. The first transaction causes a block to be authored. The second transaction does not cause a block to be authored because of this logic.

But that means the second transaction will sit in the pool until the next heartbeat block comes which will generally be a long time or possibly never if heartbeats are disabled. I think better behavior is that when a transaction arrives and it is too soon to author another block, a timer should be started so that the next block will be authored as soon the minimum block time runs out.

JoshOrndorff commented on client/consensus/manual-seal/src/heartbeat_stream.rs:
```diff
@@ -0,0 +1,100 @@
+
+pub struct HeartbeatStream<Hash> {
+	pool_stream: Box<dyn Stream<Item = EngineCommand<Hash>> + Unpin + Send>,
+	delay: Delay,
```
What is this delay field for? Is it different than the heartbeat period and min block period from the `HeartbeatOptions` struct?

seunlanlege commented on client/consensus/manual-seal/src/heartbeat_stream.rs:
```diff
@@ -0,0 +1,109 @@
+
+				// reset the timer and delay future
+				hbs.delay_future = tokio::time::delay_for(Duration::from_secs(hbs.opts.timeout));
+				hbs.last_heartbeat = Some(Instant::now());
```
This throws away the engine command yeilded by pool_stream

seunlanlege commented on client/consensus/manual-seal/src/heartbeat_stream.rs:
```diff
@@ -0,0 +1,109 @@
+					return Poll::Ready(Some(EngineCommand::SealNewBlock {
+						create_empty: true, // heartbeat blocks are empty by definition
+						finalize: hbs.opts.finalize,
+						parent_hash: None, // QUESTION: no parent hash here? Is this block still conneected with the whole chain?
```
parent_hash will be supplied by authorship task

JoshOrndorff replied:
When you put `None` it uses the best known block.

seunlanlege commented on client/consensus/manual-seal/src/heartbeat_stream.rs:
```diff
@@ -0,0 +1,109 @@
+		let mut hbs = self.get_mut();
+		match hbs.pool_stream.poll_next_unpin(cx) {
+			Poll::Ready(Some(ec)) => {
+				if let Some(last_heartbeat) = hbs.last_heartbeat {
+					// If the last heartbeat happened within min_blocktime time, we want to wait at least
+					//   until `min_blocktime` has passed.
+					if Instant::now() - last_heartbeat < Duration::from_secs(hbs.opts.min_blocktime) {
+						// We set `delay_future` here so those txs arrived after heartbeats doesn't have to wait
+						//   for `timeout`s to get processed, but only `min_blocktime`s.
+						hbs.delay_future = tokio::time::delay_for(Duration::from_secs(hbs.opts.min_blocktime));
+						return Poll::Pending;
+					}
+				}
```
This throws away the engine command yeilded by pool_stream

seunlanlege commented on client/consensus/manual-seal/src/lib.rs:
```diff
@@ -156,6 +170,9 @@ pub async fn run_instant_seal<B, CB, E, C, A, SC, T>(
 	inherent_data_providers: InherentDataProviders,
+	finalize: bool,
+	heartbeat_opts: Option<HeartbeatOptions>,
+	_replies_sender: Option<mpsc::Sender<()>>,
```
unused param

jimmychu0807 commented:
@seunlanlege I believe I will need your help from here 🙏

1. About your comment that `Some(ec)` will be thrown away, you are right. But how do I just push back the ec into the head of `pool_stream`? 

2. I have done my best on writing the tests. Please review the last two tests. I am particularly uncomfortable how the last test is written (the test on verifying min_blocktime).

If you want, feel free to commit on top of this PR. Thank you!

jimmychu0807 reviewed (commented):
@bkchr  @kianenigma Please let me know what you think of the PR. Thanks.

JoshOrndorff commented on client/consensus/manual-seal/src/lib.rs:
```diff
@@ -19,7 +19,13 @@
 //! A manual sealing engine: the engine listens for rpc calls to seal blocks and create forks.
 //! This is suitable for a testing environment.
 
+#![allow(dead_code)]
```
What is this for?

jimmychu0807 replied:
Because manual-seal is not used by default. If we don't have this line of code, compiler will warn us the code is there but not being used.

kianenigma replied:
FYI pubic items can be unused in a library.

JoshOrndorff replied:
I removed this line. It was actually hiding a valuable warning that `rpc::ManualSeal` was no longer public which is also fixed.

kianenigma commented on client/consensus/manual-seal/src/finalize_block.rs:
```diff
@@ -1,18 +1,20 @@
-// Copyright 2019 Parity Technologies (UK) Ltd.
```
was there any purpose that we have license change in the PR? It seems the old code was using the old format, maybe it was for a reason?

jimmychu0807 replied:
I find this seems to be the latest way of license. It also updating the year to 2020

kianenigma commented on client/consensus/manual-seal/src/heartbeat_stream.rs:
```diff
@@ -0,0 +1,133 @@
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+//! HeartbeatStream type and HeartbeatOptions used in instant seal
```
```suggestion
//! HeartbeatStream type and HeartbeatOptions used in instant seal

```

kianenigma replied:
+ this is a good spot to document how the whole heartbeat process is working.

kianenigma commented on client/consensus/manual-seal/src/heartbeat_stream.rs:
```diff
@@ -0,0 +1,133 @@
+}
+
+/// HeartbeatStream is combining transaction pool notification stream and generate a new block
+///   when a certain time has passed without any transactions.
```
```suggestion
/// when a certain time has passed without any transactions.
```

kianenigma commented on client/consensus/manual-seal/src/heartbeat_stream.rs:
```diff
@@ -0,0 +1,133 @@
+	/// To remember when the last block is generated
+	last_blocktime: Option<Instant>,
+	/// Heartbeat options
+	opts: HeartbeatOptions,
```
```suggestion
	options: HeartbeatOptions,
```

(opinionated)
I'd keep all variable names un-abbreviated except special 1 letter cases.

jimmychu0807 replied:
updated

kianenigma commented on client/consensus/manual-seal/src/heartbeat_stream.rs:
```diff
@@ -0,0 +1,133 @@
+}
+
+impl<Hash> Stream for HeartbeatStream<Hash> {
+
```
```suggestion
```

kianenigma reviewed (commented):
So the `min_block_time` is raising a question for me. 

With instant seal, it makes sense. Basically it makes it almost-instant-seal.

But, if I run manual seal, and receive some transactions immediately, and send the RPC command immediately, all before reaching the new `min_block_time`, will it still be delayed?

kianenigma commented on client/consensus/manual-seal/src/heartbeat_stream.rs:
```diff
@@ -0,0 +1,133 @@
+
+use crate::rpc::{EngineCommand};
+
+/// Constants for generating default heartbeat options
```
```suggestion
// Constants for generating default heartbeat options
```

This is being applied to the first one and not the rest if `///`. Besides since it is not public rustdoc `///` is not needed and `//` will suffice.

bkchr commented on client/consensus/manual-seal/src/seal_new_block.rs:
```diff
@@ -1,18 +1,20 @@
 // This file is part of Substrate.
 
-// Substrate is free software: you can redistribute it and/or modify
+// Copyright (C) 2020 Parity Technologies (UK) Ltd.
```
```suggestion
// Copyright (C) 2019 Parity Technologies (UK) Ltd.
```

bkchr commented on client/consensus/manual-seal/Cargo.toml:
```diff
@@ -20,7 +21,7 @@ jsonrpc-derive = "14.2.1"
 parking_lot = "0.10.0"
 serde = { version = "1.0", features=["derive"] }
-assert_matches = "1.3.0"
+tokio = { version = "0.2", features=["full"] }
```
PLease revert this change.

We use `futures-timer` for this.

bkchr commented on client/consensus/manual-seal/src/heartbeat_stream.rs:
```diff
@@ -0,0 +1,133 @@
+	prelude::*,
+	task::{Context, Poll},
+};
+use tokio::time::{Duration, Instant, Delay};
```
Please use from futures-timer.

bkchr commented on client/consensus/manual-seal/src/heartbeat_stream.rs:
```diff
@@ -0,0 +1,133 @@
+};
+use tokio::time::{Duration, Instant, Delay};
+
+use crate::rpc::{EngineCommand};
```
```suggestion
use crate::rpc::EngineCommand;
```

bkchr commented on client/consensus/manual-seal/src/heartbeat_stream.rs:
```diff
@@ -0,0 +1,133 @@
+use crate::rpc::{EngineCommand};
+
+/// Constants for generating default heartbeat options
+const DEFAULT_MAX_BLOCKTIME: u64 = 30;
```
Do we really need them as constants?

JoshOrndorff replied:
They won't be used anywhere else outside of this file. Should they just be variables? Or just hard-coded into the `new` function?

JoshOrndorff replied:
I've moved them to the `default` function.

bkchr commented on client/consensus/manual-seal/src/heartbeat_stream.rs:
```diff
@@ -0,0 +1,133 @@
+impl<Hash> HeartbeatStream<Hash> {
+	pub fn new(
+		pool_stream: Box<dyn Stream<Item = EngineCommand<Hash>> + Unpin + Send>,
+		opts: HeartbeatOptions
```
```suggestion
		opts: HeartbeatOptions,
```

bkchr commented on client/consensus/manual-seal/src/heartbeat_stream.rs:
```diff
@@ -0,0 +1,133 @@
+		opts: HeartbeatOptions
+	) -> Self {
+		if opts.min_blocktime > opts.max_blocktime {
+			panic!("Heartbeat options `min_blocktime` value must not be larger than `max_blocktime` value.");
```
Nope! No panics!

Return an `Error`.

jimmychu0807 replied:
I returned the error upstream. But the upstream caller is an async function of `run_instant_seal`. Can I use an `expect` there, or else how do I return a `Result` further up in an async function?

bkchr commented on client/consensus/manual-seal/src/heartbeat_stream.rs:
```diff
@@ -0,0 +1,133 @@
+			pool_stream,
+			delay_future: tokio::time::delay_for(Duration::from_secs(opts.max_blocktime)),
+			last_blocktime: None,
+			opts
```
```suggestion
			opts,
```

bkchr commented on client/consensus/manual-seal/src/heartbeat_stream.rs:
```diff
@@ -0,0 +1,133 @@
+						//   block created is passed.
+						hbs.delay_future = tokio::time::delay_for(Duration::from_secs(hbs.opts.min_blocktime)
+							- passed_blocktime);
+						return Poll::Pending;
```
This will not work. When you create a new future, you need to poll it, so it is registered to wake up the future.

jimmychu0807 replied:
The latest commit should fix this issue.

bkchr commented on client/consensus/manual-seal/src/heartbeat_stream.rs:
```diff
@@ -0,0 +1,133 @@
+			Poll::Ready(Some(ec)) => {
+				if let Some(last_blocktime) = hbs.last_blocktime {
+					// If the last block happened less than the min_blocktime time, we want to wait till
+					//   `min_blocktime` has lapsed.
```
```suggestion
					//  `min_blocktime` has lapsed.
```

bkchr commented on client/consensus/manual-seal/src/heartbeat_stream.rs:
```diff
@@ -0,0 +1,133 @@
+
+	type Item = EngineCommand<Hash>;
+
+	fn poll_next(self: Pin<&mut Self>, cx: &mut Context) -> Poll<Option<Self::Item>> {
```
The logic in here is confusing.

When I understood it correctly, you always only want to generate a block at a rate of `min_blocktime` and at maximum `max_blocktime` between blocks.

So what I would propose the following:

1. You always poll `delay_future` first. 
2. If the `delay_future` fired, check if the `pool_stream` had any events, if there was an event build a new block.
3. If there was no event, check if the `maximum_block` time passed and build a block if so.

`delay_future` should be changed to an interval to make it easier and you should also introduce an interval for the maximum block time.

bkchr commented on client/consensus/manual-seal/src/lib.rs:
```diff
@@ -19,7 +19,13 @@
 
+#![allow(dead_code)]
+
+use std::{sync::Arc, marker::{PhantomData, Send}};
```
```suggestion
use std::{sync::Arc, marker::PhantomData};
```

`Send` is in prelude.

bkchr commented on client/consensus/manual-seal/src/lib.rs:
```diff
@@ -171,286 +184,29 @@ pub async fn run_instant_seal<B, CB, E, C, A, SC, T>(
 			}
 		});
 
+	let stream: Box<dyn Stream<Item=EngineCommand<<B as BlockT>::Hash>> + Unpin + Send> =
```
Does it really require the full type here?

JoshOrndorff replied:
No. I've reduced it to `Box<dyn Stream<Item = _> + Unpin>`. I don't think it can be reduced further. This line may become obsolete after we implement your other changes anyway.

JoshOrndorff commented:
I think these changes should be made to manual seal, so that they affect both the manual seal, and instant seal which uses manual seal internally. So @kianenigma, yes I imagine the minimum block time affecting the manual seal the exact same way. Any particular node hat doesn't want that behavior can turn it off by setting the delay to zero.

# Timing Daigrams
I made some timing diagrams that I think will help us all understand the way this is supposed to work, which should make it easier to review the actual code. The horizontal axis is time. The pink marks above the axis indicate a "trigger" which means an RPC call in the case of manual seal or a transaction in the case of instant seal. The white marks below the axis indicate when a block is authored.

## Before this PR

Simple, a block is authored with every trigger. If there are no triggers, there are no blocks. If there are a bunch of really fast triggers, there are a bunch of really fast blocks (I was able to get over 1000 blocks/sec). This is how it should still work when both the min and max time are disabled.
![image](https://user-images.githubusercontent.com/2915325/92549285-e9002000-f226-11ea-9462-1c4dec047248.png)

## Max Block Time (aka "heartbeat blocks")

If the specified time elapses with no triggers, a heartbeat block is authored anyway. This helps users and offchain tools know the chain is still live. In the figure the max block time is five time units.

![image](https://user-images.githubusercontent.com/2915325/92549446-41cfb880-f227-11ea-97ef-83bec413920f.png)

## Min Block Time

Introduces a "cool down" period after each block is authored in which no other blocks can be authored. If any triggers arrive during the cool down period, they are suppressed, and a single block is authored after the cooldown period. This single block ensures that at least one block is authored after each trigger, which is particularly useful in instant seal where it will clear any transactions that have arrived in the pool. The cooldown periods are shown just below the axis, and are 1 time unit long.

![image](https://user-images.githubusercontent.com/2915325/92550052-ae978280-f228-11ea-9db5-8fbe03fb71c7.png)

# How many timers?

> you should also introduce an interval for the maximum block time.

@bkchr, I initially also thought there should be two times. One for max block time, one for min block time. Jimmy changed my mind when he pointed out that there will always be exactly one block scheduled at any point in time. If a trigger arrives and a block is authored, then you schedule a block at the maximum time, replacing any previously scheduled block. If a trigger arrives and no block is authored because the min time has not yet elapsed, then you schedule a block at the minimum block time from the last authored block, replacing any previously scheduled block.

Now that I've written this out, I think I actually think I like the two-timer design better.

JoshOrndorff commented:
I've changed my mind. I no longer think the min_block_time should be applied to manual seal. When using manual seal you are able to manually specify which block to use as the parent. Enforcing a minimum block time means suppressing rpc calls that come too close together and replacing them by a single authoring event later. If the RPC calls specify different parents, then we would somehow have to choose which parent to author on. That sounds like a mess.

So my new opinion is that **heartbeat blocks should be available to both instant- and manual- seal, but min-block-time should only be available to instant-seal**.

jimmychu0807 commented:
My thought is keeping block authoring and finalizing in manual-seal totally manual, aka sending a RPC instruction to author block, and another RPC instruction to finalize block. So it is as dumb as it can be. Then we do all the smart logic in the instant-seal. So I prefer to keep the `max_blocktime` and `cooldown_blocktime` (formerly called `min_blocktime`, but I found `cooldown_blocktime` convey its meaning better) in instant-seal section.

jimmychu0807 commented:
@bkchr @seunlanlege 

1. I have updated the `heartbeat` period and `cooldown` period code. Any of them can be `None`. If both are `None`, then it is just the normal instant-seal.

2. I have fixed the delay future that is not being polled. Tested it.

3. I am updating the license mainly for the year to 2020. The text is mostly copied from `frame/benchmark/src`. Let me know if you have strong opinion against them.

4. I commented two lines in the `tests.rs` (mentioned in the code) because it didn't pass when I run it. Seun, does it pass when you run? Anyway, I think it can be fixed in another PR.

Please review them and let me know if the PR is good. Thank you! 🙏

jimmychu0807 commented on client/consensus/manual-seal/src/tests.rs:
```diff
@@ -0,0 +1,429 @@
+
+	// Ensure the forked block is in the db
+	// note: The following line fails the test
+	// let imported = rx2.await.unwrap().unwrap();
```
@seunlanlege This is the line that fail when I run the test. Does it pass for you?

JoshOrndorff replied:
When I was working on #7070 I noticed that one of the already-existing tests was failing. Is that the case here Jimmy? Did this test fail before you made changes?

jimmychu0807 replied:
This is the test that fail 4/5 times approximately. Sometimes it pass...

gnunicorn commented:
@jimmychu0807 please merge master

jimmychu0807 commented:
will do @gnunicorn

jimmychu0807 commented:
I think this PR is good for merge after getting the latest master @gnunicorn

@bkchr you have previously commented, want to see if you still have any concern.
@seunlanlege @andresilva  Seun, you are the original author of manual-seal and Andre you oversee consensus, let me know if you have any comment on the work.

Thank you for all your time 🙏

bkchr commented on client/consensus/manual-seal/src/heartbeat_stream.rs:
```diff
@@ -0,0 +1,149 @@
+
+	fn poll_next(mut self: Pin<&mut Self>, cx: &mut Context) -> Poll<Option<Self::Item>> {
+		match self.command_stream.poll_next_unpin(cx) {
+
```
```suggestion
```

bkchr commented on client/consensus/manual-seal/src/heartbeat_stream.rs:
```diff
@@ -0,0 +1,149 @@
+					},
+					None => Poll::Pending,
+				}
+			}, // End of `Poll::Pending`
+
```
```suggestion
			},
```

bkchr commented on client/consensus/manual-seal/src/heartbeat_stream.rs:
```diff
@@ -0,0 +1,149 @@
+impl<Hash> Stream for HeartbeatStream<Hash> {
+	type Item = EngineCommand<Hash>;
+
+	fn poll_next(mut self: Pin<&mut Self>, cx: &mut Context) -> Poll<Option<Self::Item>> {
```
The logic in this function is too complicated in here.

There are also some questions that arise, what happens when we drop such an engine command? Like we should finalize a block and now it is just dropped?

And why didn't you followed my earlier proposal for this? That would not result in dropping any events.

bkchr commented:
The logic of the heartbeat stream is still too complicated and it drops events. I assume that we should not drop events, because they could be rather important?

jimmychu0807 commented on client/consensus/manual-seal/src/heartbeat_stream.rs:
```diff
@@ -0,0 +1,149 @@
+						//      |------time_passed-------+------wait_further------|
+						//   last blocktime           current                  cooldown
+						//
+						let time_passed = Instant::now() - last_blocktime;
```
Updated to use `saturating_duration_since` and `checked_sub`

jimmychu0807 commented:
![IMG_2977](https://user-images.githubusercontent.com/898091/96855761-c2710e00-148f-11eb-887e-36f58e8fb195.jpg)

@bkchr 
- Just want to confirm the understanding with you. The current implementation handles the above 3 cases ( (2), (3), and (4) in the diagram) - this is also conveying the same idea as Joshy's diagram way above. With your suggested implementation, does it handle (3) in a simpler manner? (I can't seem to find a way to do it without similar complexity as current one).

- Also on your raised concern that events are dropped: events are dropped only in case (2), when consecutive ECs come in a burst. But it doesn't matter, because new block has already been scheduled to be sealed after the cooldown. ECs only have two events only, one to seal unfinalized block, another to seal finalized block.

bkchr commented on client/consensus/manual-seal/src/heartbeat_stream.rs:
```diff
@@ -0,0 +1,149 @@
+	heartbeat: Option<Duration>,
+	// The cooldown duration after a block has been authored.
+	cooldown: Option<Duration>,
+	// Whether the created block is finalized of not.
```
```suggestion
	// Whether the created block is finalized.
```

bkchr commented:
>     * Just want to confirm the understanding with you. The current implementation handles the above 3 cases ( (2), (3), and (4) in the diagram) - this is also conveying the same idea as [Joshy's diagram way above](https://github.com/paritytech/substrate/pull/6951#issuecomment-689274697). With your suggested implementation, does it handle (3) in a simpler manner? (I can't seem to find a way to do it without similar complexity as current one).

I still think that the implementation of the function could be made simpler, by simpler I mean better to read. Currently it is a huge match which requires some thinking to understand what is going on.

> 
>     * Also on your raised concern that events are dropped: events are dropped only in case (2), when consecutive ECs come in a burst. But it doesn't matter, because new block has already been scheduled to be sealed after the cooldown. ECs only have two events only, one to seal block, another to finalize block.

We still would drop the finalize block event when it comes in a cooldown, which would be a weird behavior. If the user sends this event and we just drop it.

JoshOrndorff commented:
> We still would drop the finalize block event when it comes in a cooldown, which would be a weird behavior. If the user sends this event and we just drop it.

This would be addressed by something like https://github.com/paritytech/substrate/pull/7070

gnunicorn commented:
Closing due to it being stale for quite a while.
